<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Introduction to BASH and Sequence Analysis - 9&nbsp; SNP Calling and Genome Assembly</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./10_File_Format.html" rel="next">
<link href="./8_seq_to_Alignment.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./9_SNP_Calling_and_Genome_Assembly.html"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">SNP Calling and Genome Assembly</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to BASH and Sequence Analysis</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Installing VS Code and FileZilla</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./First_Time_Accessing_the_Server.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">First Time Accessing the Server</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./1_Introducing_the_Shell.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introducing the Shell</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./2_Navigating_Files_and_Directories.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Navigating Files and Directories</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./3_Working_With_Files_and_Directories.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Working With Files and Directories</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./4_Pipes_and_Filters.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Pipes and Filters</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./5_loops.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Loops</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./6_How_2_copy.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">How do you copy to and from the remote computer?</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./7_bioinformatics_workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">A typical bioinformatics workflow</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./8_seq_to_Alignment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">From Sequencing to Genome Alignment</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./9_SNP_Calling_and_Genome_Assembly.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">SNP Calling and Genome Assembly</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10_File_Format.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">File Format</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#snp-calling" id="toc-snp-calling" class="nav-link active" data-scroll-target="#snp-calling"><span class="header-section-number">9.1</span> SNP Calling</a>
  <ul class="collapse">
  <li><a href="#indexing-the-sorted-alignment" id="toc-indexing-the-sorted-alignment" class="nav-link" data-scroll-target="#indexing-the-sorted-alignment"><span class="header-section-number">9.1.1</span> Indexing the sorted alignment</a></li>
  <li><a href="#freebayes-snp-calling" id="toc-freebayes-snp-calling" class="nav-link" data-scroll-target="#freebayes-snp-calling"><span class="header-section-number">9.1.2</span> <code>freebayes</code> SNP calling</a></li>
  <li><a href="#bcftools-snp-calling" id="toc-bcftools-snp-calling" class="nav-link" data-scroll-target="#bcftools-snp-calling"><span class="header-section-number">9.1.3</span> <code>bcftools</code> SNP calling</a></li>
  <li><a href="#and-compare-the-results" id="toc-and-compare-the-results" class="nav-link" data-scroll-target="#and-compare-the-results"><span class="header-section-number">9.1.4</span> And compare the results!</a></li>
  </ul></li>
  <li><a href="#genome-assembly" id="toc-genome-assembly" class="nav-link" data-scroll-target="#genome-assembly"><span class="header-section-number">9.2</span> Genome Assembly</a>
  <ul class="collapse">
  <li><a href="#why-fragment-the-genome-understanding-the-need-for-short-reads-in-genome-assembly" id="toc-why-fragment-the-genome-understanding-the-need-for-short-reads-in-genome-assembly" class="nav-link" data-scroll-target="#why-fragment-the-genome-understanding-the-need-for-short-reads-in-genome-assembly"><span class="header-section-number">9.2.1</span> Why Fragment the Genome? Understanding the Need for Short Reads in Genome Assembly</a></li>
  <li><a href="#types-of-genome-assembly" id="toc-types-of-genome-assembly" class="nav-link" data-scroll-target="#types-of-genome-assembly"><span class="header-section-number">9.2.2</span> Types of Genome Assembly</a></li>
  </ul></li>
  <li><a href="#de-novo-assembly-of-chloroplast-genome" id="toc-de-novo-assembly-of-chloroplast-genome" class="nav-link" data-scroll-target="#de-novo-assembly-of-chloroplast-genome"><span class="header-section-number">9.3</span> De-novo assembly of chloroplast genome</a>
  <ul class="collapse">
  <li><a href="#preparing-the-reads-fastq-files" id="toc-preparing-the-reads-fastq-files" class="nav-link" data-scroll-target="#preparing-the-reads-fastq-files"><span class="header-section-number">9.3.1</span> Preparing the reads (fastq files)</a></li>
  </ul></li>
  <li><a href="#velvet-de-novo-genome-assembly" id="toc-velvet-de-novo-genome-assembly" class="nav-link" data-scroll-target="#velvet-de-novo-genome-assembly"><span class="header-section-number">9.4</span> Velvet de-novo genome assembly</a>
  <ul class="collapse">
  <li><a href="#velvet-output-files" id="toc-velvet-output-files" class="nav-link" data-scroll-target="#velvet-output-files"><span class="header-section-number">9.4.1</span> Velvet output files</a></li>
  </ul></li>
  <li><a href="#spades-de-novo-genome-assembly" id="toc-spades-de-novo-genome-assembly" class="nav-link" data-scroll-target="#spades-de-novo-genome-assembly"><span class="header-section-number">9.5</span> SPAdes de novo genome assembly</a>
  <ul class="collapse">
  <li><a href="#spades-output-files" id="toc-spades-output-files" class="nav-link" data-scroll-target="#spades-output-files"><span class="header-section-number">9.5.1</span> SPAdes output files</a></li>
  </ul></li>
  <li><a href="#stats-using-seqkit" id="toc-stats-using-seqkit" class="nav-link" data-scroll-target="#stats-using-seqkit"><span class="header-section-number">9.6</span> Stats using <code>seqkit</code> !</a></li>
  <li><a href="#alignment-using-mummer" id="toc-alignment-using-mummer" class="nav-link" data-scroll-target="#alignment-using-mummer"><span class="header-section-number">9.7</span> Alignment using MUMmer</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">SNP Calling and Genome Assembly</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="snp-calling" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="snp-calling"><span class="header-section-number">9.1</span> SNP Calling</h2>
<p><strong>SNP calling</strong> (Single Nucleotide Polymorphism calling) is a process in genomics used to identify single nucleotide variants between sequences. These are the most common type of genetic variation among people, plants, animals, and other organisms. SNPs represent a difference in a single nucleotide (A, T, C, or G) in the DNA sequence among individuals or between a sample and a reference genome.</p>
<p>For SNP calling we can use <code>freebayes</code> or <code>samtools</code>.</p>
<p>We can call SNPs from read alignments - why not reuse the ones from previous section (i.e.&nbsp;<code>Alignments_sorted.bam</code>)?</p>
<section id="indexing-the-sorted-alignment" class="level3" data-number="9.1.1">
<h3 data-number="9.1.1" class="anchored" data-anchor-id="indexing-the-sorted-alignment"><span class="header-section-number">9.1.1</span> Indexing the sorted alignment</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> myresults</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index Alignments_sorted.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="freebayes-snp-calling" class="level3" data-number="9.1.2">
<h3 data-number="9.1.2" class="anchored" data-anchor-id="freebayes-snp-calling"><span class="header-section-number">9.1.2</span> <code>freebayes</code> SNP calling</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">freebayes</span> <span class="at">--fasta-reference</span> najas.fasta <span class="at">-v</span> UNFILTERED_RESULTS.vcf Alignments_sorted.bam</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--vcf</span> UNFILTERED_RESULTS.vcf <span class="at">--recode</span> <span class="at">--out</span> freeB_SNPs <span class="at">--minQ</span> 15 <span class="at">--minDP</span> 10</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="bcftools-snp-calling" class="level3" data-number="9.1.3">
<h3 data-number="9.1.3" class="anchored" data-anchor-id="bcftools-snp-calling"><span class="header-section-number">9.1.3</span> <code>bcftools</code> SNP calling</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> mpileup <span class="at">-f</span> najas.fasta Alignments_sorted.bam <span class="kw">|</span> <span class="ex">bcftools</span> call <span class="at">-mv</span> <span class="at">-Ov</span> <span class="at">-o</span> bcftools_calls.vcf</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--vcf</span> bcftools_calls.vcf <span class="at">--recode</span> <span class="at">--out</span> bcftools_filtered_callsSNPs <span class="at">--minQ</span> 15 <span class="at">--minDP</span> 10</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="and-compare-the-results" class="level3" data-number="9.1.4">
<h3 data-number="9.1.4" class="anchored" data-anchor-id="and-compare-the-results"><span class="header-section-number">9.1.4</span> And compare the results!</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-a</span> bcftools_filtered_callsSNPs.recode.vcf <span class="at">-b</span> freeB_SNPs.recode.vcf <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="genome-assembly" class="level2" data-number="9.2">
<h2 data-number="9.2" class="anchored" data-anchor-id="genome-assembly"><span class="header-section-number">9.2</span> Genome Assembly</h2>
<p>Genome assembly refers to the process of taking a large number of short DNA sequences and putting them back together to create a representation of the original chromosomes from which the DNA originated.</p>
<p><img src="img/genome_assembly.png" class="img-fluid"></p>
<section id="why-fragment-the-genome-understanding-the-need-for-short-reads-in-genome-assembly" class="level3" data-number="9.2.1">
<h3 data-number="9.2.1" class="anchored" data-anchor-id="why-fragment-the-genome-understanding-the-need-for-short-reads-in-genome-assembly"><span class="header-section-number">9.2.1</span> Why Fragment the Genome? Understanding the Need for Short Reads in Genome Assembly</h3>
<p>Cutting the genome into short reads before sequencing is a necessity due to current technological limitations.</p>
<ul>
<li><p><strong>Sequencer Capabilities:</strong> Most high-throughput sequencing technologies, such as Illumina, PacBio, and Oxford Nanopore, have physical and technical limitations on the length of DNA they can accurately read. Short-read technologies like Illumina can only sequence fragments up to a few hundred base pairs long, while long-read technologies like PacBio and Oxford Nanopore can read longer fragments (thousands to tens of thousands of base pairs), but still not entire chromosomes or genomes in one go.</p></li>
<li><p><strong>Accuracy and Error Rates:</strong> Sequencing shorter fragments tends to produce more accurate and reliable data. Long reads often have higher error rates, which can complicate the assembly process. Combining short, accurate reads with long reads can improve the overall quality of the genome assembly.</p></li>
</ul>
</section>
<section id="types-of-genome-assembly" class="level3" data-number="9.2.2">
<h3 data-number="9.2.2" class="anchored" data-anchor-id="types-of-genome-assembly"><span class="header-section-number">9.2.2</span> Types of Genome Assembly</h3>
<p><strong>Reference-based Assembly:</strong></p>
<ul>
<li>This method uses an existing reference genome from a related organism as a template to guide the assembly of the new genome. It aligns the short DNA reads to the reference genome and fills in the gaps.</li>
<li><code>Advantages:</code> Faster and less computationally intensive, as it relies on the structure and sequence of the reference genome.</li>
<li><code>Disadvantages:</code> May not accurately assemble regions that differ significantly from the reference genome, potentially missing novel sequences or structural variations.</li>
</ul>
<p><strong>De-novo Genome Assembly:</strong></p>
<ul>
<li>This method assembles the genome from scratch, without the use of a reference genome. It involves overlapping short reads to form longer contiguous sequences (contigs) and then ordering and orienting these contigs to create scaffolds representing the complete genome.</li>
<li><code>Advantages:</code> Capable of revealing novel sequences, structural variations, and accurately assembling genomes of organisms with no close reference.</li>
<li><code>Disadvantages:</code> Computationally intensive and requires higher coverage (more reads) to ensure accuracy and completeness.</li>
</ul>
</section>
</section>
<section id="de-novo-assembly-of-chloroplast-genome" class="level2" data-number="9.3">
<h2 data-number="9.3" class="anchored" data-anchor-id="de-novo-assembly-of-chloroplast-genome"><span class="header-section-number">9.3</span> De-novo assembly of chloroplast genome</h2>
<p>We want to assemble our reads to get a new chloroplast genome assembly we’ll use the reads that aligned to the reference to save some time.</p>
<section id="preparing-the-reads-fastq-files" class="level3" data-number="9.3.1">
<h3 data-number="9.3.1" class="anchored" data-anchor-id="preparing-the-reads-fastq-files"><span class="header-section-number">9.3.1</span> Preparing the reads (fastq files)</h3>
<p><strong>Remove ‘unmapped’ reads</strong></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-h</span> <span class="at">-F</span> 4 Alignments_sorted.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>performs the following operation using <code>samtools</code>:</p>
<ol type="1">
<li><p><code>view</code>: This is a <code>samtools</code> command used to view or extract alignments from a SAM or BAM file.</p></li>
<li><p><code>-h</code>: The <code>-h</code> option specifies that the output should include the SAM header information. It means that the header lines, which provide metadata about the SAM/BAM file, will be included in the output.</p></li>
<li><p><code>-F 4</code>: The <code>-F</code> option is used to exclude specific flag values from the output. In this case, <code>-F 4</code> indicates that reads with the “unmapped” flag (bit 3) will be excluded from the output. This means that only mapped reads will be included in the output.</p></li>
<li><p><code>Alignments_sorted.bam</code>: This is the input BAM file from which the reads will be extracted. The path and file name of the BAM file are provided as the last argument.</p></li>
</ol>
<p>Overall, the command will extract mapped reads from the <code>Alignments_sorted.bam</code> BAM file and display them in SAM format, including the header information. The output will exclude any reads that are marked as “unmapped” in the BAM file.</p>
<p><strong>Create paired-end FASTQ files from the sam/bam file</strong></p>
<p><code>samtools fastq input.sam -1 First.fastq -2 Second.fastq</code></p>
<p>performs the following operation using <code>samtools</code>:</p>
<ol type="1">
<li><p><code>fastq</code>: This is a <code>samtools</code> command used to convert a SAM/BAM file to FASTQ format, which is a standard format for storing sequencing reads and their quality scores.</p></li>
<li><p><code>input.sam</code>: This is the input SAM file that you want to convert to FASTQ format. Replace <code>input.sam</code> with the path and filename of the actual SAM file you want to convert.</p></li>
<li><p><code>-1 First.fastq</code>: The <code>-1</code> option specifies the output file for the first mates (R1) of paired-end reads. The converted FASTQ data for the first mates will be saved in a file named <code>First.fastq</code>. Replace <code>First.fastq</code> with the desired filename.</p></li>
<li><p><code>-2 Second.fastq</code>: The <code>-2</code> option specifies the output file for the second mates (R2) of paired-end reads. The converted FASTQ data for the second mates will be saved in a file named <code>Second.fastq</code>. Replace <code>Second.fastq</code> with the desired filename.</p></li>
</ol>
<p>Overall, the command will convert the alignments in the <code>input.sam</code> SAM file to paired-end FASTQ files (<code>First.fastq</code> and <code>Second.fastq</code>), allowing you to work with the sequencing reads in a standard and widely used format for downstream analysis, such as read mapping, variant calling, or quality assessment. Note that if your input file is in BAM format instead of SAM, you can use the same <code>samtools fastq</code> command, and <code>samtools</code> will automatically handle the conversion from BAM to FASTQ.</p>
<p>To get ALL aligning reads:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> myresults/</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-h</span>  <span class="at">-F</span> 4 Alignments_sorted.bam <span class="kw">|</span> <span class="ex">samtools</span> fastq <span class="at">-1</span> first.fastq <span class="at">-2</span> second.fastq</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>To get only reads aligning in proper pairs:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> myresults/</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-h</span>  <span class="at">-f</span> 3 Alignments_sorted.bam <span class="kw">|</span> <span class="ex">samtools</span> fastq <span class="at">-1</span> first.fastq <span class="at">-2</span> second.fastq</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now you have paired-end properly aligned FASTQ files <code>first.fastq</code> and <code>second.fastq</code> which are redy for assembly. We will use two de-novo assembler: <strong>Velvet</strong> and <strong>SPAdes</strong> to do the assembly and compare the results.</p>
</section>
</section>
<section id="velvet-de-novo-genome-assembly" class="level2" data-number="9.4">
<h2 data-number="9.4" class="anchored" data-anchor-id="velvet-de-novo-genome-assembly"><span class="header-section-number">9.4</span> Velvet de-novo genome assembly</h2>
<p>https://github.com/dzerbino/velvet</p>
<p><strong>setting up a Velvet assembly using <code>velveth</code></strong></p>
<p><code>velveth</code> program, is part of the Velvet de novo genome assembly software.</p>
<p>The following command is for setting up a Velvet assembly(to create the hash table for the Velvet assembler) with the specified k-mer size (29 in this case, you can try a random one 21 - 141 so we can compare) using paired-end short read data provided in separate FASTQ files (<code>first.fastq</code> and <code>second.fastq</code>).</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">velveth</span> kmer29_velvet_Assembly 29 <span class="at">-shortPaired</span> <span class="at">-fastq</span> <span class="at">-separate</span> first.fastq second.fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Let’s break down the components of the command:</p>
<ol type="1">
<li><p><code>velveth</code>: This is the command to invoke the <code>velveth</code> program for setting up the Velvet assembly.</p></li>
<li><p><code>kmer29_velvet_Assembly</code>: This is the name of the output directory where the Velvet assembly files will be stored. The assembly will be created in a folder named <code>kmer29_velvet_Assembly</code>.</p></li>
<li><p><code>29</code>: This is the k-mer size parameter, which specifies the length of the overlapping subsequences used in the assembly process. In this case, the k-mer size is set to 29.</p></li>
<li><p><code>-shortPaired</code>: This option indicates that the input reads are in paired-end format.</p></li>
<li><p><code>-fastq</code>: This option specifies that the input files are in FASTQ format.</p></li>
<li><p><code>-separate</code>: This option indicates that the paired-end reads are provided in separate FASTQ files. The two input FASTQ files (<code>first.fastq</code> and <code>second.fastq</code>) contain the forward and reverse reads of the paired-end data.</p></li>
</ol>
<p>The resulting assembly files will be stored in the <code>kmer29_velvet_Assembly</code> directory. The chosen k-mer size can significantly impact the quality and performance of the assembly process.</p>
<p><strong><code>velvetg</code> Assembly</strong></p>
<p>The following command will do the gemome assembly using <code>velvetg</code></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">velvetg</span> kmer29_velvet_Assembly <span class="at">-clean</span> yes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The command <code>velvetg kmer29_velvet_Assembly -clean yes</code> is used to run the Velvet genome assembler with specific options on an assembly created using a k-mer size of 29.</p>
<p>Explanation of the components:</p>
<ul>
<li><code>velvetg</code>: This is the command to run the Velvet genome assembler.</li>
<li><code>kmer29_velvet_Assembly</code>: This specifies the name of the assembly directory generated using a k-mer size of 29. It refers to the output directory where Velvet has previously stored the intermediate files and results of the assembly process.</li>
<li><code>-clean yes</code>: This is an option for Velvet that instructs the assembler to remove intermediate files generated during the assembly process. By specifying <code>-clean yes</code>, the command will clean up and delete these intermediate files once the assembly process is complete.</li>
</ul>
<p>In summary, the command is running Velvet on an assembly that was previously created using a k-mer size of 29 and then cleaning up the intermediate files after the assembly is done.</p>
<section id="velvet-output-files" class="level3" data-number="9.4.1">
<h3 data-number="9.4.1" class="anchored" data-anchor-id="velvet-output-files"><span class="header-section-number">9.4.1</span> Velvet output files</h3>
<p>The main output of Velveth is a set of files that represent the input data in a format optimized for the subsequent assembly step performed by Velvetg. The most important output files are <code>contigs.fa</code> and <code>stats.txt</code>.</p>
<ol type="1">
<li><p><strong>contigs.fa (or contigs.fasta):</strong> This file contains the assembled contigs generated by Velveth during the initial de novo assembly process. Contigs are contiguous sequences that represent overlapping subsequences (k-mers) of the input reads. Each contig represents a portion of the genome that Velveth was able to assemble based on the input sequencing data. The <code>contigs.fa</code> file is in FASTA format, with each contig represented by a sequence header (name) and its corresponding nucleotide sequence.</p></li>
<li><p><strong>stats.txt:</strong> This file provides statistical information about the assembly process and the resulting contigs. The <code>stats.txt</code> file typically includes metrics such as the total number of contigs, the N50 length (a measure of assembly quality), the total assembly size, and various other statistics related to contig lengths and coverage. This information is essential for assessing the quality and completeness of the assembly.</p></li>
</ol>
</section>
</section>
<section id="spades-de-novo-genome-assembly" class="level2" data-number="9.5">
<h2 data-number="9.5" class="anchored" data-anchor-id="spades-de-novo-genome-assembly"><span class="header-section-number">9.5</span> SPAdes de novo genome assembly</h2>
<p>https://github.com/ablab/spades SPAdes is a widely used bioinformatics software tool for performing de novo genome assembly from short-read sequencing data. The name “SPAdes” stands for “St.&nbsp;Petersburg Genome Assembler.” The following command will perform de novo assembly of paired-end short read data with a specified k-mer size of 29 using SPAdes.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">spades.py</span> <span class="at">-o</span> kmer29_Spades_assembly <span class="at">-1</span> first.fastq <span class="at">-2</span> second.fastq <span class="at">-k</span> 29</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Let’s break down the components of the command:</p>
<ol type="1">
<li><p><code>spades.py</code>: This is the command to run the SPAdes genome assembler.</p></li>
<li><p><code>-o kmer29_Spades_assembly</code>: The <code>-o</code> option specifies the output directory where the assembled contigs and related files will be stored. The directory named <code>kmer29_Spades_assembly</code> will be created to hold the assembly results.</p></li>
<li><p><code>-1 first.fastq</code>: The <code>-1</code> option indicates the input file containing the forward reads (R1) of the paired-end data, which is stored in the <code>first.fastq</code> file.</p></li>
<li><p><code>-2 second.fastq</code>: The <code>-2</code> option indicates the input file containing the reverse reads (R2) of the paired-end data, which is stored in the <code>second.fastq</code> file.</p></li>
<li><p><code>-k 29</code>: The <code>-k</code> option specifies the k-mer size used during the assembly. In this case, a k-mer size of 29 is used. The k-mer size determines the length of the overlapping subsequences used in the assembly process.</p></li>
</ol>
<p>SPAdes is a widely used bioinformatics tool for de novo genome assembly from short-read sequencing data. It employs multiple algorithms to handle various aspects of the assembly process, including k-mer-based assembly and de Bruijn graph construction. The <code>-k</code> option allows users to specify different k-mer sizes to optimize the assembly based on the characteristics of the input data. The assembled contigs and scaffolds will be saved in the output directory specified with the <code>-o</code> option for further analysis and exploration.</p>
<section id="spades-output-files" class="level3" data-number="9.5.1">
<h3 data-number="9.5.1" class="anchored" data-anchor-id="spades-output-files"><span class="header-section-number">9.5.1</span> SPAdes output files</h3>
<p>SPAdes generates multiple output files as a result of the genome assembly process. These files contain valuable information about the assembled genome and related statistics. Below are the main output files typically produced by SPAdes:</p>
<ol type="1">
<li><p><strong>contigs.fasta:</strong> This file contains the assembled contigs, which are contiguous sequences of the genome without gaps. Each contig represents a portion of the genome that SPAdes was able to assemble based on the input sequencing data.</p></li>
<li><p><strong>scaffolds.fasta:</strong> The scaffolds file contains longer sequences, known as scaffolds, which are formed by connecting contigs using paired-end read information. Scaffolding helps to bridge gaps between contigs and creates more extended sequences.</p></li>
<li><p><strong>contigs.paths:</strong> This file provides the contig paths in the de Bruijn graph used during the assembly process. It shows how the contigs were assembled and the order of overlapping k-mers.</p></li>
<li><p><strong>scaffolds.paths:</strong> Similar to contigs.paths, this file provides the paths of scaffolds in the de Bruijn graph. It shows the order in which scaffolds were connected to form the final assembly.</p></li>
<li><p><strong>assembly_graph.fastg:</strong> This file contains the assembly graph in FASTG format, which represents the relationships between contigs and scaffolds. It provides a graphical representation of how the genome was assembled.</p></li>
<li><p><strong>before_rr.fasta:</strong> If the input data includes long reads (e.g., PacBio or Oxford Nanopore), this file contains contigs generated from these reads before the read error correction step.</p></li>
<li><p><strong>corrected.fasta:</strong> If the input data includes long reads, this file contains corrected contigs after the read error correction step.</p></li>
<li><p><strong>dataset.info:</strong> This file provides information about the input datasets used in the assembly, such as the number of reads, read length, and other statistics.</p></li>
<li><p><strong>params.txt:</strong> This file contains the parameters and settings used in the SPAdes assembly.</p></li>
<li><p><strong>spades.log:</strong> The log file records information about the assembly process, including warnings, errors, and resource usage.</p></li>
</ol>
<p>Additionally, SPAdes generates various subdirectories within the output directory that contain intermediate files and data produced during the assembly process. These files are used internally by SPAdes and are essential for its assembly algorithm.</p>
<p>The specific set of output files and their contents may vary depending on the input data, the assembly options chosen, and the version of SPAdes used. The <code>contigs.fasta</code> and <code>scaffolds.fasta</code> files are typically the most important output files as they represent the final assembled sequences of the genome. We can use these output files for downstream analyses, such as gene prediction, functional annotation, and comparative genomics studies.</p>
</section>
</section>
<section id="stats-using-seqkit" class="level2" data-number="9.6">
<h2 data-number="9.6" class="anchored" data-anchor-id="stats-using-seqkit"><span class="header-section-number">9.6</span> Stats using <code>seqkit</code> !</h2>
<p>Now let’s look at stats using seqkit.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># stats for velvet assembly</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="ex">seqkit</span> stats <span class="at">-a</span> kmer29_velvet_Assembly/contigs.fa</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co">#stats for SPAdes assembly</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="ex">seqkit</span> stats <span class="at">-a</span> kmer29_Spades_assembly/contigs.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>performs the following operation using the <code>seqkit</code> tool:</p>
<ol type="1">
<li><p><code>seqkit</code>: This is the command to run the <code>seqkit</code> tool, a bioinformatics utility for working with biological sequence data.</p></li>
<li><p><code>stats</code>: This is one of the functionalities provided by <code>seqkit</code>. The <code>stats</code> subcommand is used to generate various statistics for the sequences in the input file.</p></li>
<li><p><code>-a</code>: The <code>-a</code> option specifies that the statistics should be calculated for each sequence in the input file (in this case, <code>contigs.fasta</code>) in addition to the overall statistics for the whole file.</p></li>
<li><p><code>kmer29_Spades_assembly/contigs.fasta</code>: This is the input file that contains the assembled contigs in FASTA format. The path and filename point to the <code>contigs.fasta</code> file generated by the SPAdes assembly with k-mer size 29.</p></li>
</ol>
<p>The output of the command will display various statistics for the sequences in the <code>contigs.fasta</code> file. These statistics commonly include:</p>
<ol type="1">
<li><p><strong>Total Sequences:</strong> The total number of sequences (contigs) present in the input file.</p></li>
<li><p><strong>Total Bases:</strong> The total number of nucleotide bases in all the sequences combined.</p></li>
<li><p><strong>Minimum Length:</strong> The length of the shortest sequence (contig) in the file.</p></li>
<li><p><strong>Maximum Length:</strong> The length of the longest sequence (contig) in the file.</p></li>
<li><p><strong>Mean Length:</strong> The average length of all the sequences in the file.</p></li>
<li><p><strong>N50 Length:</strong> A measure of assembly quality, which represents the length at which half of the total bases are found in sequences that are equal to or longer than this length. N50 is used as an indicator of the contiguity of the assembly.</p></li>
<li><p><strong>GC Content:</strong> The percentage of G (guanine) and C (cytosine) nucleotides in all the sequences combined.</p></li>
<li><p><strong>Sequence Composition:</strong> A summary of the number of sequences with specific GC content ranges (e.g., 20-30%, 30-40%, etc.).</p></li>
</ol>
<p>These statistics are valuable for evaluating the quality of the genome assembly and for understanding the characteristics of the assembled contigs. A high N50 value indicates good contig contiguity, while the GC content provides insights into the nucleotide composition of the genome.</p>
<p>By analyzing the <code>seqkit stats</code> output, researchers can assess the performance of the assembly and determine the next steps in their genomic analysis, such as gene prediction, functional annotation, and comparative genomics.</p>
</section>
<section id="alignment-using-mummer" class="level2" data-number="9.7">
<h2 data-number="9.7" class="anchored" data-anchor-id="alignment-using-mummer"><span class="header-section-number">9.7</span> Alignment using MUMmer</h2>
<p>MUMmer (MUltiple MUMs for DNA Sequence) is a bioinformatics software package designed for the rapid alignment of large DNA sequences. It is particularly useful for aligning whole genomes and is capable of handling sequences ranging from small bacterial genomes to large eukaryotic genomes.</p>
<p>We will align our de-novo assembled genome against the reference genome we have (i.e.&nbsp;thalassia.fasta or najas.fasta). First,</p>
<p><strong>Align our de-novo assembled genome against the reference genome</strong></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> myresults</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co">#teach-maker2</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> /mnt/s-ws/everyone/scie4002_refs/thalassia.fasta .</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co">#teach-maker3</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> /mnt/s-ws/everyone/thalassia.fasta .</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> kmer29_Spades_assembly</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="ex">nucmer</span> ../thalassia.fasta contigs.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Here’s a breakdown of the components of the command:</p>
<ol type="1">
<li><p><code>nucmer</code>: This is the command to run the NUCmer program.</p></li>
<li><p><code>../thalassia.fasta</code>: This is the path and filename of the first input FASTA file, which contains the reference genome or sequence. In this case, the reference sequence is stored in the file <code>thalassia.fasta</code> located in the directory <code>myresults</code>.</p></li>
<li><p><code>contigs.fasta</code>: This is the path and filename of the second input FASTA file, which contains the assembled contigs or sequences that you want to compare to the reference genome. In this case our de-novo assembled genome using SPAdes.</p></li>
</ol>
<p>NUCmer performs whole-genome pairwise alignment by identifying maximal unique matches (MUMs) between the two input sequences. It then constructs an alignment between these MUMs to identify conserved regions and variations between the reference genome and the assembled contigs.</p>
<p>The output of the NUCmer program a file <code>out.delta</code>:</p>
<ul>
<li><code>out.delta</code>: This is the main output file containing the sequence alignment in delta format, which represents the location and size of the MUMs.</li>
</ul>
<p><strong>concatenate the contigs into a big fasta file</strong> using MUMmer. We could just concatenate our sequences, but that will result in an ugly output we can use MUMmer tools to order and orient our contigs MUMMER uses a reference sequence.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="ex">show-tiling</span> <span class="at">-p</span> Spades_concatenated_contigs.fasta out.delta </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">nucmer</span> ../thalassia.fasta Spades_concatenated_contigs.fasta </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="ex">mummerplot</span> <span class="at">-t</span> png out.delta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>there’ll be an error due to a bug, we can fix this though</strong></p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> mouse out.gp <span class="op">&gt;</span> out_fixed.gp</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="ex">gnuplot</span> out_fixed.gp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./8_seq_to_Alignment.html" class="pagination-link" aria-label="From Sequencing to Genome Alignment">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">From Sequencing to Genome Alignment</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./10_File_Format.html" class="pagination-link" aria-label="File Format">
        <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">File Format</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>